### ISSUE TITLE
Support needed - Microphone sensitivity and distance issue

### ISSUE THREAD

**Issue by KhArtNJava** [alphacep/vosk-api#117]
Hello.

I am trying to make something like smart home with voice recognition subsystem.
My current pocketsphinx solution is recognizing my words with custom dictionary very well from 2-3 meters from mic. Yeah, doesn't hear me sometimes, even from 15-30 cm from mic, but it's not a big problem.

By Nikolay Shmyrev's suggestion I've tried vosk with test_microphone.py and vosk-model-small-ru-0.4 - speech recognition results are impressive! But, only from 30 cm from mic. If I move for 1.5 meters from mic, it doesn't recognize my speech. I also switched on ventilation in my toilet and bath rooms, to make some background noise.

My test machine is Win 10, Python 3.8 with pyaudio.

I've not bad mic Dexp U700, with it I can hear in my headphones, speech sound from smartphone's speaker (that is not in loud mode), that is located in another room. So, mic has very good sensitivity.

How can vosk be configured to work in common apartments environment, with some noises in background?

**Comment by nshmyrev** [comment]
Share those audio recordings from your microphone, I'll take a look.

**Comment by KhArtNJava** [comment]
[test1.zip](https://github.com/alphacep/vosk-api/files/4708113/test1.zip)
Here is an example, that I recorded with Audacity with enabled vosk in background.
Vosk didn't recognize anything from this record.

**Comment by KhArtNJava** [comment]
Also, 
python test_simple.py untitled.wav
doesn't recognize anything from this record (Signed 16-bit PCM). 
But this record (30 cm from mic) it recognizes very well:
[untitled2.zip](https://github.com/alphacep/vosk-api/files/4708122/untitled2.zip)

**Comment by nshmyrev** [comment]
Well, I'll take a look closer but the audio is obviously not so good. If you really want to have a smart home you'd better buy something like this: https://www.seeedstudio.com/ReSpeaker-Mic-Array-v2-0.html

**Comment by KhArtNJava** [comment]
OKay, I found it on AliExpress. Interesting demo at page https://www.seeedstudio.com/ReSpeaker-Mic-Array-v2-0.html , but it's really expensive for hobby - about 7010 roubles at AliExpress :). Yep, I can afford it, but if I'll cut another expenses. Do you know, are there any other cheaper solutions? ))) Thanks.

**Comment by nshmyrev** [comment]
If you have RPi you can get much cheaper older model https://www.seeedstudio.com/ReSpeaker-4-Mic-Array-for-Raspberry-Pi.html, but overall the latest model is the only one which has hardware echo cancellation, i.e. which can play music and in the same time record clean speech. Others can't do that easily.

**Comment by KhArtNJava** [comment]
Oh, I saw it on Ali too, but I don't see in docs, that it supports voice extraction.
Does it really support Voice Extraction?

**Comment by nshmyrev** [comment]
It has voice extraction features but they are probably not that simple to setup. You can check here for example https://github.com/voice-engine/ec

**Comment by KhArtNJava** [comment]
I am trying to understand how it works - https://wiki.seeedstudio.com/ReSpeaker_Mic_Array_v2.0/ for Voice Extraction.
They are recording audio from v.2.0 device and in result they have 6 channel audio wav file. So, to use it, I should use 6_channels_firmware.bin firmware and channel 0, that has Extracted Voice, right? I still has not looked at https://github.com/voice-engine/ec , but if it so easy with v.2.0 device, I probably should buy it, as the simplest solution.

**Comment by KhArtNJava** [comment]
Николай, прошу прощения за оффтоп - посоветуйте, пожалуйста, TTS русский. Сейчас остановил свой выбор на https://github.com/Olga-Yakovleva/RHVoice , лучше не нашёл.

**Comment by nshmyrev** [comment]
Хороших нет вроде пока. Несколько команд делают, но результата я ещё не видел. Можете попробовать https://github.com/alphacep/tn2-wg, но там GPU карта нужна.

**Comment by KhArtNJava** [comment]
Николай, можно вас попросить выложить пример записи синтеза от https://github.com/alphacep/tn2-wg ? Оно русский язык держит? Выбор голосов имеется? Увидел там только одну модель, непонятно, для русского языка ли и какой голос.
Будьте так любезны, запишите, пожалуйста, синтез следующих предложений, всё в виде одного аудиофайла'
"Привет, Юлиана, как дела? Сегодня погода теплая, за окном плюс 25 градусов тепла".
У меня нет Nvidia карты, думаю, стоит ли приобретать её ради этого.

**Comment by nshmyrev** [comment]
> Николай, можно вас попросить выложить пример записи синтеза от https://github.com/alphacep/tn2-wg ? 

[test.zip](https://github.com/alphacep/vosk-api/files/4841395/test.zip)

> Оно русский язык держит? 

Да, но над качеством надо ещё работать

> Выбор голосов имеется?

Нет, там только один голос.

> У меня нет Nvidia карты, думаю, стоит ли приобретать её ради этого.

В скором времени будет работать на процессоре, уже есть такие разработки. Для своих улучшений карты понадобятся и не одна.

**Comment by KhArtNJava** [comment]
Благодарю Вас за оперативность.
Прослушал так же и https://github.com/alphacep/tn2-wg/blob/master/test.wav, что-то как-то не впечатлило. В обоих примерах - в присланном вами и в репозитории неправильно ставит ударения, голос неживой. У вас в описании сказано "Tacotron2 + Waveglow Russian". Глянул так же глянул примеры на странице https://nv-adlr.github.io/WaveGlow, в разделе Tacotron 2 + WaveGlow - небо и земля. Англоязычная озвучка, например, про Scientists... крайне реалистична. От чего зависит такая большая разница? Один и тот же движок, получается, на английском говорит идеально, а на русском - как робот?

**Comment by nshmyrev** [comment]
На английском тоже далеко от идеала. Над ударениями надо работать, когда-нибудь сделаем.

**Comment by KhArtNJava** [comment]
А то, что голос менее сочный, не живой, по сравнению с английским?
 Самая лучшая озвучка у них первая. Остальные "отдают электроникой", но даже с учётом этого выглядит более реалистично, чем имеющаяся русская озвучка.
Интересно, когда будет русская версия хотя бы на схожем уровне, как сейчас у английской версии?

**Comment by KhArtNJava** [comment]
Николай, наконец-то пришла игрушка ReSpeaker_Mic_Array_v2.0 - почти 2 месяца из Китая ехала, с сайта seeedstudio.
Так вот, что-то как-то она меня разочаровала, Я ожидал большего от неё, а именно АРУ по голосу.
Возможно, что готовить её ещё не научился правильно...
Сразу по инструкции прошил на 6_channels_firmware.bin по инструкции https://wiki.seeedstudio.com/ReSpeaker_Mic_Array_v2.0/#update-firmware .
После чего на компе записал с помощью Audacity 6 канал:
![image](https://user-images.githubusercontent.com/757085/89526803-fd4c9980-d7f0-11ea-9ea3-5218a6e01ab3.png)
Запись:
 - сначала рядом с компом стою;
 - отхожу в коридор, примерно на 2 метра от компа;
 - отхожу ещё дальше в коридор;
 - в конце коридора произношу фразу "Я в коридоре".

Что-то как-то грустно работает их Voice Extraction. Я ожидал, что будет АРУ для выделения голоса, который слышен из далека.
[Temp.zip](https://github.com/alphacep/vosk-api/files/5034523/Temp.zip)

Посоветуйте, пожалуйста, по поводу Voice auto volume gain...

**Comment by nshmyrev** [comment]
> Посоветуйте, пожалуйста, по поводу Voice auto volume gain...

AGC включён как написано в [инструкции](https://wiki.seeedstudio.com/ReSpeaker_Mic_Array_v2.0/#tuning)?

python tuning.py AGCONOFF 0

**Comment by KhArtNJava** [comment]
А точно нужно 0 ставить в 
python tuning.py AGCONOFF 0 ?
Оно же вроде как выключает?

**Comment by KhArtNJava** [comment]
Пробовал и с 1 и с 0 
AGCONOFF: 0
AGCONOFF: 1

Всё равно голос не вытягивает на желаемый уровень...

**Comment by KhArtNJava** [comment]
Вот примеры записей 
[with 0 value.zip](https://github.com/alphacep/vosk-api/files/5034773/with.0.value.zip)
[with 1.zip](https://github.com/alphacep/vosk-api/files/5034774/with.1.zip)

**Comment by nshmyrev** [comment]
Там ещё есть параметры у AGC - https://wiki.seeedstudio.com/ReSpeaker_Mic_Array_v2.0/#faq

**Comment by KhArtNJava** [comment]
Поигрался с настройками.
Включил шумодавы и AGC, но всё равно Vosk не распознает толком...
```
c:\Python38\usb_4_mic_array>python tuning.py -r
name                     value
-------------------------------
tuning.py:181: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.
  print('{:24} {}'.format(name, dev.read(name)))
AECFREEZEONOFF           0
AECNORM                  2.0
AECPATHCHANGE            0
AECSILENCELEVEL          9.99999993922529e-09
AECSILENCEMODE           1
AGCDESIREDLEVEL          0.009999999776482582
AGCGAIN                  13.588893339037895
AGCMAXGAIN               1000.0
AGCONOFF                 1
AGCTIME                  0.9824014389887452
CNIONOFF                 1
DOAANGLE                 17
ECHOONOFF                1
FREEZEONOFF              0
FSBPATHCHANGE            0
FSBUPDATED               0
GAMMAVAD_SR              15.0
GAMMA_E                  1.0
GAMMA_ENL                1.0
GAMMA_ETAIL              1.0
GAMMA_NN                 1.100000023841858
GAMMA_NN_SR              1.100000023841858
GAMMA_NS                 1.0
GAMMA_NS_SR              1.0
HPFONOFF                 1
MIN_NN                   0.30000001192092896
MIN_NN_SR                0.30000001192092896
MIN_NS                   0.15000000596046448
MIN_NS_SR                0.15000000596046448
NLAEC_MODE               0
NLATTENONOFF             0
NONSTATNOISEONOFF        1
NONSTATNOISEONOFF_SR     1
RT60                     0.8999999761581421
RT60ONOFF                1
SPEECHDETECTED           0
STATNOISEONOFF           1
STATNOISEONOFF_SR        1
TRANSIENTONOFF           1
VOICEACTIVITY            0
```

Похоже, что зря выкинул деньги на этот микрофон... Чудо не произошло... 

Пример записи с этими настройками
[output2.zip](https://github.com/alphacep/vosk-api/files/5035259/output2.zip)

**Comment by KhArtNJava** [comment]
```
c:\Temp\vosk-api\python\example>python test_simple.py output2.wav
LOG (VoskAPI:ConfigureV2():build-libkaldiwrap/model.cc:202) Decoding params beam=10 max-active=3000 lattice-beam=2
LOG (VoskAPI:ConfigureV2():build-libkaldiwrap/model.cc:205) Silence phones 1:2:3:4:5:6:7:8:9:10
LOG (VoskAPI:ComputeDerivedVars():ivector-extractor.cc:183) Computing derived variables for iVector extractor
LOG (VoskAPI:ComputeDerivedVars():ivector-extractor.cc:204) Done.
LOG (VoskAPI:RemoveOrphanNodes():nnet-nnet.cc:948) Removed 1 orphan nodes.
LOG (VoskAPI:RemoveOrphanComponents():nnet-nnet.cc:847) Removing 2 orphan components.
LOG (VoskAPI:Collapse():nnet-utils.cc:1472) Added 1 components, removed 2
LOG (VoskAPI:CompileLooped():nnet-compile-looped.cc:345) Spent 0.0559669 seconds in looped compilation.
LOG (VoskAPI:ReadDataFiles():build-libkaldiwrap/model.cc:259) Loading HCL and G from model/graph/HCLr.fst model/graph/Gr.fst
LOG (VoskAPI:ReadDataFiles():build-libkaldiwrap/model.cc:281) Loading winfo model/graph/phones/word_boundary.int
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : "раз"
}
{
  "partial" : "раз два"
}
{
  "partial" : "раз два"
}
{
  "partial" : "раз два"
}
{
  "partial" : "раз два"
}
{
  "result" : [{
      "conf" : 1.000000,
      "end" : 1.410000,
      "start" : 1.080000,
      "word" : "раз"
    }, {
      "conf" : 1.000000,
      "end" : 1.680000,
      "start" : 1.410000,
      "word" : "два"
    }],
  "text" : "раз два"
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "text" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "text" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "text" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "text" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "text" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "partial" : ""
}
{
  "text" : ""
}
```

**Comment by nshmyrev** [comment]
CNIONOFF                 1

я бы отрубил

**Comment by KhArtNJava** [comment]
Выключил CNIONOFF , всё равно не распознает...

**Comment by nshmyrev** [comment]
Выложите запись. Попробуйте говорить сразу издалека.

**Comment by KhArtNJava** [comment]
Поднял HPFONOFF  до 2 - чуть- чуть почище стал звук, но всё равно не распознает. Сейчас попробую издалека сразу.
[output2.zip](https://github.com/alphacep/vosk-api/files/5035440/output2.zip)

**Comment by KhArtNJava** [comment]
Запись сразу из коридора:
[output2.zip](https://github.com/alphacep/vosk-api/files/5035450/output2.zip)


```
c:\Python38\usb_4_mic_array>python tuning.py -r
name                     value
-------------------------------
tuning.py:181: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.
  print('{:24} {}'.format(name, dev.read(name)))
AECFREEZEONOFF           0
AECNORM                  2.0
AECPATHCHANGE            0
AECSILENCELEVEL          9.99999993922529e-09
AECSILENCEMODE           1
AGCDESIREDLEVEL          0.009999999776482582
AGCGAIN                  20.012782901525497
AGCMAXGAIN               1000.0
AGCONOFF                 1
AGCTIME                  0.9824014389887452
CNIONOFF                 0
DOAANGLE                 143
ECHOONOFF                1
FREEZEONOFF              0
FSBPATHCHANGE            0
FSBUPDATED               0
GAMMAVAD_SR              15.0
GAMMA_E                  1.0
GAMMA_ENL                1.0
GAMMA_ETAIL              1.0
GAMMA_NN                 1.100000023841858
GAMMA_NN_SR              1.100000023841858
GAMMA_NS                 1.0
GAMMA_NS_SR              1.0
HPFONOFF                 2
MIN_NN                   0.30000001192092896
MIN_NN_SR                0.30000001192092896
MIN_NS                   0.15000000596046448
MIN_NS_SR                0.15000000596046448
NLAEC_MODE               0
NLATTENONOFF             0
NONSTATNOISEONOFF        1
NONSTATNOISEONOFF_SR     1
RT60                     0.8999999761581421
RT60ONOFF                1
SPEECHDETECTED           0
STATNOISEONOFF           1
STATNOISEONOFF_SR        1
TRANSIENTONOFF           1
VOICEACTIVITY            0
```

**Comment by nshmyrev** [comment]
Шумновата запись. Что это шумит всё-таки?

**Comment by KhArtNJava** [comment]
В комнате стоит вентилятор, но его шум почти полностью срезается HPFONOFF = 2. Плюс на кухне работает стиральная машина.
Суть проекта состоит в том, чтобы речь нормально распознавалась именно в бытовых условиях.

**Comment by nshmyrev** [comment]
Начать надо сначала с задачи попроще. Вытянуть звук в тишине из коридора. Хотя это вряд ли получится, слишком большое эхо.

**Comment by KhArtNJava** [comment]
Я тут ещё экспериментировал с настройками, что аж пришлось перешиваться, чтобы на дефолтовые настройки сброситься...
В общем, игрушка эта годится максимум для определения угла, откуда звук идёт.
Приспособлю в дальнейшем для робота, чтобы голосом можно было управлять им.
Для серьёзных нужд, кроме как конференций, сия платка не годится.
Дело просто в том, что простой эксперимент не осилила она:
 - в одном углу комнаты поставил ReSpeaker_Mic_Array_v2.0 , рядом лежит телефон с Google Keyboard с определением Voice to text;
 - Я стою в метре от них, произношу фразу "Компьютер, включи свет на кухне" - обе распознают фразу на 100%;
 - поворачиваюсь к ним спиной, произношу ту же фразу - сотка фразу распознаёт, ReSpeaker + VOSK - ничего не распознают из этой фразы;
 - отхожу в другой конец комнаты, встаю лицом к ним, опять та же фраза - обе определяют её как "включи свет на кухне" (пропускают слово Компьютер);
 - разворачиваюсь спиной к ним, стоя в том же месте - сотка определяет "включи свет на кухне", пропуская слово Компьютер в очередной раз, ReSpeaker + VOSK - ничего не распознают из этой фразы.

Клавиатура гугла на обычном телефоне со слабеньким микрофоном "сделала" аж 4 микрофона ReSpeaker. 

Пищаль (((. 

Николай, есть идеи, какие ещё могут быть решения для выделения речи в шуме?

**Comment by gittitov** [comment]
Nikolay, I also have ReSpeaker_Mic_Array_v2.0
It worked very poorly compared to android-vosk app.
Then I took example/test_microphone.py tag 0.3.15 - it became very good!

Ubuntu 20.04. pyenv Python 3.7.9 on two different computers with Intel and AMD

**Comment by nshmyrev** [comment]
Yes, new reworked api though sounddevice should be much better than before. I recommend 0.3.17, it is faster too.

**Comment by gittitov** [comment]
in my case 0.3.17 is much worse. Permanent recognition errors

**Comment by gittitov** [comment]
i use 0.3.17 but example/test_microphone.py from 0.3.15

**Comment by nshmyrev** [comment]
We have released Russian model 0.22 with far-field training data for better far-field training. Not much to fix here otherwise.
